### DISCUSSION TITLE
Snes (Demon's Crest) SPC outputs wrong song

### DISCUSSION THREAD

**Discussion by ArcadeTV** [vgmtrans/vgmtrans#221] [General]
I don't know if it's a bug or improper ripped SPC but this is happening:

When I play song 05 from
http://snesmusic.org/v2/profile.php?set&selected=668
it's the correct song.

When I use the same SPC in VGMtrans or CAPSPC then this song is magically changed into another song. This happens to some other SPC in the soundtrack as well.

**Comment by loveemu** [comment]
Try resaving SPC file by using SNES SPC700 Player (save file after the player starts playback). In most cases, the wrong song issue happens on the SPC dumped before finishing the loading.

## Long answer

In some games, the sound program can store multiple sequences to the RAM, and there are no obvious way to know what sequence is currently playing. VGMTrans will try grabbing the song index by heuristic analysis. (from song number variable, or sometimes from the pointers of sequence parser, etc.)

Unfortunately, the result can be wrong sometimes:

* Song previously played is incorrectly detected.
    - SPC file was dumped before the song gets fully-loaded.
* Jingles or SFXs can't be detected.
    - Heuristic routines often ignore them.

If VGMTrans loads all the sequences existing in the RAM, the problem will be gone. Instead, you need to find the target song you want, from the bunch of unnamed sequences. Also, there's no way to make a VGMColl manually from VGMInstrSet and VGMSeq... right?

**Comment by ViLXDRYAD** [comment]
first of all, thanks you a lot for this amazing software i have found very useful! :D 

tried the fix above without luck, it was for the theme 22 - Dr. Light theme from Megaman X, http://spc.joshw.info/m/Mega%20Man%20X%20%5bRockman%20X%5d%20(1993-12-17)(Capcom).7z when it converts the track to midi, it output the 11 - Chill Penguin theme instead

also unrelated BUT i wonder why some midi ends a bit later than its supposed? like the 16 - Storm Eagle theme from the same game

thanks you a lot for everything! x3

**Comment by kylesacher** [comment]
I've run into this issue with Dancing Mad (http://snesmusic.org/v2/profile.php?selected=14287&profile=set) and made unsuccessful attempts to fix it. ff6-314a.spc (Dancing Mad Part 1), ff6-314b.spc (Dancing Mad Part 2), and ff6-314c.spc (Dancing Mad Part 3) play correctly in other SPC players, but VGMTrans treats all three of them as Part 1.

If I copy the first 512 bytes (x100 through x2FF in an external hex editor) from 314a into 314b, I can make WinAmp play Part 1 from 314b, instead of Part 2 as it was before. I know the SPC-700 treats these first 512 bytes differently from the rest of memory, but I'm not sure what VGMTrans does with it.

VGMTrans appears to know the entire region to load, based on its hex viewer, but it's ignoring everything starting from x20AD (x21AD in an external hex editor). I believe that's the start of the first track for Part 2, and all three parts are in all three files. VGMTrans doesn't seem to be looking at the first 512 bytes to decide where to find tracks, though. Instead, I'm seeing in the code (AkaoSnesScanner.cpp, line157) that it's looking for a bit pattern that will always tell it that the header starts at x1C00, and that header points to Part 1 in all three files.

**Comment by KungFuFurby** [comment]
ARAM map analysis would be required to better resolve this issue, and it is not always stable across all of the known builds. Unfortunately, I mostly haven't built any of them yet, though for N-SPC I do have a default ARAM map that is mostly compatible across a good portion of the variants... but not always.
